### PR TITLE
Support and test mypy 0.910, 0.920, and 0.921

### PR DESCRIPTION
**Pull request by christianbundy**
## Change Summary

Problem: We aren't testing mypy version compatibility, which makes it hard to know when we're making/breaking compatibility. In my local testing, it seems that we've recently broken mypy 0.910 compatibility.

Solution: Add mypy version matrix to CI and fix compatibility regressions.

## Related issue number

Link: https://github.com/samuelcolvin/pydantic/discussions/3538#discussioncomment-1867912

See-also: #3175 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by christianbundy** [comment]
please review

**Comment by samuelcolvin** [comment]
Thanks for this, I'm stealing some ideas for #3572.

File: pydantic/mypy.py [hunk 1]
```diff
@@ -357,17 +363,23 @@ def add_construct_method(self, fields: List['PydanticModelField']) -> None:
...
-        self_type = TypeVarType(self_tvar_name, tvar_fullname, -1, [], obj_type)
+        tvd = TypeVarDef(self_tvar_name, tvar_fullname, -1, [], obj_type)
         self_tvar_expr = TypeVarExpr(self_tvar_name, tvar_fullname, [], obj_type)
         ctx.cls.info.names[self_tvar_name] = SymbolTableNode(MDEF, self_tvar_expr)
 
+        # Backward-compatible with TypeVarDef from Mypy 0.910.
+        if isinstance(tvd, TypeVarType):  # pragma: no cover
+            self_type = tvd
+        else:
+            self_type = TypeVarType(tvd)  # type: ignore[call-arg]
```
Comment (samuelcolvin) [inline pydantic/mypy.py:374]: I'm a bit confused here, you define `self_type` but don't use it.

Is this a mistake?

Comment (samuelcolvin) [reply pydantic/mypy.py:374]: don't worry, I'm being dumb, ignore me.

**Pull request closed without merging**